modules/lang-mustache/src/main/java/org/elasticsearch/action/search/template/AbstractSearchTemplateAction.java

```diff
@@ -0,0 +1,88 @@
+ */
+
+package org.elasticsearch.action.search.template;
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.ActionResponse;
+import org.elasticsearch.action.support.ActionFilters;
+import org.elasticsearch.action.support.HandledTransportAction;
+import org.elasticsearch.cluster.metadata.IndexNameExpressionResolver;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.script.ExecutableScript;
+import org.elasticsearch.script.Script;
+import org.elasticsearch.script.ScriptService;
+import org.elasticsearch.script.mustache.MustacheScriptEngineService;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportService;
+
+import java.util.function.Supplier;
+
+import static java.util.Collections.emptyMap;
+import static org.elasticsearch.script.ScriptContext.Standard.SEARCH;
+
+public abstract class AbstractSearchTemplateAction<Request extends AbstractSearchTemplateRequest<Request>, Response extends ActionResponse>
+        extends HandledTransportAction<Request, Response> {
+
+    private static final String TEMPLATE_LANG = MustacheScriptEngineService.NAME;
+
+    private final ClusterService clusterService;
+    private final ScriptService scriptService;
+
+    public AbstractSearchTemplateAction(Settings settings, String actionName, ThreadPool threadPool, TransportService transportService,
+                                        ActionFilters actionFilters, IndexNameExpressionResolver indexNameExpressionResolver,
+                                        Supplier<Request> request, ClusterService clusterService, ScriptService scriptService) {
+        super(settings, actionName, threadPool, transportService, actionFilters, indexNameExpressionResolver, request);
+        this.clusterService = clusterService;
+        this.scriptService = scriptService;
+    }
+
+    @Override
+    protected void doExecute(Request request, ActionListener<Response> listener) {
+        new AsyncSearchTemplateAction(request, listener).start();
+    }
+
+    /**
+     * Execute the action when the search template is rendered
+     */
+    protected abstract void doExecute(Request request, ActionListener<Response> listener, BytesReference source);
+
+    class AsyncSearchTemplateAction {
+
+        private final Request request;
+        private final ActionListener<Response> listener;
+
+        public AsyncSearchTemplateAction(Request request, ActionListener<Response> listener) {
+            this.request = request;
+            this.listener = listener;
+        }
+
+        void start() {
+            try {
+                Script script = new Script(request.getScript(), request.getScriptType(), TEMPLATE_LANG, request.getScriptParams());
+                ExecutableScript executable = scriptService.executable(script, SEARCH, emptyMap(), clusterService.state());
```

Comment:
It might be a good idea to make the ScriptContext a required constructor argument. I know both the implementations will set it to SEARCH but it'll force any extender to think about the appropriate context. If you leave it like it is they might not think about it and are much more likely to make a mistake.